src/backtrace.rs

```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use alloc::fmt::{Display, Formatter, Result};
+use core::arch::asm;
+
+pub const MAX_BACKTRACE_ADDRESSES: usize = 10;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub(crate) enum BTReturnAddress {
+    ReturnAddress(u64),
+    End,
+    InvalidFramePointer(u64),
+    InvalidReturnAddress(u64),
+}
+
+impl Display for BTReturnAddress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            BTReturnAddress::ReturnAddress(addr) => {
+                writeln!(f, "0x{addr:x}")
+            }
+            BTReturnAddress::InvalidFramePointer(addr) => {
+                writeln!(f, "Invalid frame pointer: 0x{addr:x}")
+            }
+            BTReturnAddress::InvalidReturnAddress(addr) => {
+                writeln!(f, "Null address at frame pointer: 0x{addr:x}")
+            }
+            BTReturnAddress::End => write!(f, ""),
+        }
+    }
+}
+
+pub(crate) fn backtrace() -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
+    // Safe because we are just reading a register
+    let fp = unsafe {
+        let mut tmp: u64;
+        asm!("mv {0}, x8", out(reg) tmp);
+        tmp
+    };
+
+    unsafe { backtrace_internal(fp) }
+}
+
+// Safety
+// The caller must ensure that fp is a valid address on the stack frame.
+unsafe fn backtrace_internal(fp: u64) -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
+    let mut result = [BTReturnAddress::End; MAX_BACKTRACE_ADDRESSES];
+    let mut index = 0;
+    let mut skip = 2;
+
+    // Safe because we trust the linker placed these symbols correctly.
+    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+
+    let mut fp = fp;
+    loop {
+        if fp < stack_start || fp > stack_end {
+            result[index] = BTReturnAddress::InvalidFramePointer(fp);
+            break;
+        }
+
+        // Safe because we are only reading in the valid stack trace area
+        let address = unsafe { (fp as *const u64).offset(-1).read_volatile() }; // RA/PC
+        fp = unsafe { (fp as *const u64).offset(-2).read_volatile() }; // next FP
```

Comment:
Sorry if this a duplicate, but my Github comment rendering has been really weird.

My test could be flawed, but what if fp == stack_start when we begin execution? Basically, we are effectively addressing *(fp - 16) on line 74, meaning that we can underflow the bounds of the stack?

```
fn assert_stack() {
     // Safe because we trust the linker placed these symbols correctly.
     let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
     let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
     let fp = stack_start;
     //  Same as if fp < stack_start || fp > stack_end on 67
     if fp >= stack_start && fp <= stack_end {
        let address = unsafe { (fp as *const u64).offset(-1) as u64 };
        let address = unsafe { (fp as *const u64).offset(-2) as u64 };
        assert!(address < stack_end);
        assert!(address > stack_start);
     }
}
```

Reply:
On a somewhat related note, can we treat the stack as an as `&u64` slice that begins with stack_start, and with `len == (stack_end - stack_start) / 8`? That way, we can do the unsafe{} cast once, and will get bounds check for free, instead of relying on multiple unsafe checks.

Reply:
Actually, there's no math involved, except for fp - stack_start. You can probably even mark the method as safe (but that's a separate discussion). 

Also, you simplify the code by replacing the loop with a `while index < MAX_BACKTRACE_ADDRESSES`.

For example, the following appears to be the equivalent of your code:

```
// Safety
// The caller must ensure that fp is a valid address on the stack frame.
unsafe fn backtrace_internal(fp: u64) -> [BTReturnAddress; MAX_BACKTRACE_ADDRESSES] {
    use s_mode_utils::print::CONSOLE;
    use s_mode_utils::print::*;
    let mut result = [BTReturnAddress::End; MAX_BACKTRACE_ADDRESSES];
    let mut index = 0;
    let mut skip = 2;

    // Safe because we trust the linker placed these symbols correctly.
    let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
    let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
    let mut fp  = fp;
    let stack = unsafe {core::slice::from_raw_parts(stack_start as *const u64, (stack_end - stack_start) as usize)};
    loop {  
        if fp < (stack_start+2) || fp > stack_end {
            result[index] = BTReturnAddress::InvalidFramePointer(fp);
            break;
        }

        let offset = ((fp - stack_start - 2)/8) as usize;
        if offset > 0 && let Some(address) = stack.get(offset) && let Some(current_fp) = stack.get(offset-1) {
            fp = *current_fp;
            let address = *address;     
            if address == 0 {
                result[index] = BTReturnAddress::InvalidReturnAddress(fp);
                break;
            }
            
            if skip > 0 {
                skip -= 1;
                continue;
            }
    
            result[index] = BTReturnAddress::ReturnAddress(address);
            index += 1;
    
            if index >= MAX_BACKTRACE_ADDRESSES {
                break;
            }        
        } else {
            result[index] = BTReturnAddress::InvalidReturnAddress(fp);
            break;
        }
    }

    result
}

```

Reply:
fixed